### ISSUE TITLE
Model forward with cache giving incorrect logits

### ISSUE THREAD

**Issue by Pasewark** [lucidrains/native-sparse-attention-pytorch#20]
Thanks for your amazing work on this!

I have been training some models using this code and noticed some issues when I generate with the models. I believe there is something wrong when using the Transformer forward method with cache. I specifically notice a difference in logits when generating with cache vs without. I wrote a test here to reproduce the error:

```
from native_sparse_attention_pytorch.transformer import Transformer
from native_sparse_attention_pytorch.compress_networks import GroupedMLP
import torch

dim_head=64
compress_block_size=16
num_kv_heads=2
mlp_expand_factor=.3
vocab_size=32000
hidden_size=1280
num_hidden_layers=22
num_heads=20
sliding_window_size=512
fine_block_size=16
num_fine_selected=16

compress_mlp=GroupedMLP(
        dim_head = dim_head,
        compress_block_size = compress_block_size,
        heads = num_kv_heads,
        expand_factor = mlp_expand_factor
    )
model=Transformer(
    num_tokens = vocab_size,
    dim = hidden_size,
    depth = num_hidden_layers,
    heads = num_heads,
    dim_head = dim_head,
    kv_heads = num_kv_heads,
    use_sparse_attn = True,
    use_flex_sliding_window = True,
    use_triton_fine_selection = False,
    use_flex_fine_selection = False,
    sparse_attn_kwargs = dict(
        sliding_window_size = sliding_window_size,
        compress_block_size = compress_block_size,
        compress_mlp = compress_mlp,
        selection_block_size = fine_block_size,
        num_selected_blocks = num_fine_selected,
        use_diff_topk = True,
        interpolated_importance_score = False,
        query_heads_share_selected_kv = True
    ),
).cuda()

def sample_forced(model, x, use_cache_kv=True):
    """
    Step-by-step forward pass, but always forcing the next token to be
    the same as in 'x', rather than sampling from the probabilities.

    Returns:
      forced_logits (torch.Tensor): [batch, seq_len-1, vocab_size]
      forced_true_token_logits (torch.Tensor): [batch, seq_len-1], the
          logit for the correct next token at each step.
    """
    batch_size, seq_len = x.shape
    cache = None

    all_logits = []
    true_token_logits = []

    with torch.no_grad():
        for i in range(seq_len - 1):
            out=x[:,:i+1]
            logits, next_cache = model.forward(
                out,
                cache=cache,
                return_cache=True,
                disable_flex=True,
                disable_triton_kernel=(i > 0)
            )
            if use_cache_kv:
                cache = next_cache

            # logits for predicting the *next* token
            next_logit = logits[:, -1, :]  # shape [batch, vocab_size]
            all_logits.append(next_logit)

            # gather just the logit corresponding to the *true* next token in x
            true_token_i = x[:, i + 1]  # shape [batch]
            true_token_logit_i = next_logit[torch.arange(batch_size), true_token_i]
            true_token_logits.append(true_token_logit_i)

    forced_logits = torch.stack(all_logits, dim=1)          # [batch, seq_len-1, vocab_size]
    forced_true_token_logits = torch.stack(true_token_logits, dim=1)  # [batch, seq_len-1]
    return forced_logits, forced_true_token_logits


def test_sample_equivalence(model, x, use_cache):
    """
    1) Get logits from a single forward pass on x.
    2) Get logits from forced step-by-step approach.
    3) Compare the two (they should match closely if everything is correct).
    4) Return the "true next-token" logits for each approach
       (shape [batch, seq_len-1]).
    """
    with torch.no_grad():
        # standard forward pass: shape [batch, seq_len, vocab_size]
        full_logits = model(x)  # no loss, returns logits
        # forced sampling step-by-step
        forced_logits, forced_true_token_logits = sample_forced(model, x, use_cache)

        # For single forward pass, the logit that predicts the token x[:, i+1]
        # (the "true next-token logit") is at full_logits[:, i, x[:, i+1]].
        # We'll gather those into shape [batch, seq_len-1].
        batch_size, seq_len, vocab_size = full_logits.shape
        # slice to remove the last position (no next-token for final index)
        # shape now [batch, seq_len - 1, vocab_size]
        full_logits_for_pred = full_logits[:, :-1, :]  
        # gather the correct indices: x[:, 1:] is shape [batch, seq_len-1]
        # so expand it with an extra dim for gather
        true_token_indices = x[:, 1:].unsqueeze(-1)  # [batch, seq_len-1, 1]
        # gather along the last dim -> [batch, seq_len-1, 1], then squeeze
        true_token_logits_full = torch.gather(full_logits_for_pred, 2, true_token_indices).squeeze(-1)

        # Compare forced_logits vs. full_logits (they're [batch, seq_len-1, vocab_size])
        diff_all = (full_logits[:, :-1, :] - forced_logits).abs().max()
        print("Max abs difference (all logits) =", diff_all.item())

        # Also compare the just-gathered “true token” logits
        diff_true = (true_token_logits_full - forced_true_token_logits).abs().max()
        print("Max abs difference (true token logits) =", diff_true.item())

        return true_token_logits_full, forced_true_token_logits



x = torch.randint(0, vocab_size, (1, 200)).cuda()
print('Run with use_cache=False')
ttf_full, ttf_forced=test_sample_equivalence(model,x,False) # differences less than 1e-5
print('Run with use_cache=True')
ttf_full, ttf_forced=test_sample_equivalence(model,x,True) # differences greater than .1
```

Basically this makes a random sequence of tokens and then computes logits in 2 ways. First is the regular model forward. Next it computes logits similar to generation, but forcing each generated token to match the tokens in x instead of sampling. So if everything is working correctly, the logits should be the same for each method (within numerical precision).

If the cache is not used in the forced sampling, the logit differences are very close. But if the cache is used the logit differences are not very close. The differences seem to be larger with a trained model. I noticed that my trained model has very bad generations with using cache, but without cache the generations are fine. I don't know if the issue is actually in the cache, or if it is some other problem with the forward_inference method in the attention.

**Comment by lucidrains** [comment]
@Pasewark would you like to give this a shot? very straightforward task for a budding machine learning engineer

**Comment by lucidrains** [comment]
@Pasewark discovered and resolved one issue this morning with inference

would you like to retry and see if it is any better?

**Comment by Pasewark** [comment]
I am still getting a difference when using the cache. I'll also take a look and see if I can figure out what might be wrong, but I don't make any promises that I will be able to figure it out soon.

**Comment by lucidrains** [comment]
@Pasewark thanks! it would have been lucky to have knocked it out with just one issue

will iron out the remaining discrepancy in the selection attention by end of month

**Comment by lucidrains** [comment]
@Pasewark want to try yet again?

**Comment by lucidrains** [comment]
@Pasewark you are applying this to Amazon Search?

**Comment by Pasewark** [comment]
I haven't tested out every case, but it seems like the logits are now correct when the block overlap len is 0 and fine_block_size==compress_block_size. However, when the overlap is not 0 or fine_block_size!=compress_block_size, there is still some difference. Here are the parameters I am using, in case I am doing something wrong:
```
dim_head=64
compress_block_size=16
num_kv_heads=2
mlp_expand_factor=.3
vocab_size=32000
hidden_size=1280
num_hidden_layers=22
num_heads=20
sliding_window_size=512
fine_block_size=16   # set !=compress_block_size and get different logits
num_fine_selected=16
overlap_size=0   # set to 4 and get different logits


compress_mlp=GroupedMLP(
        dim_head = dim_head,
        compress_window_size = compress_block_size + overlap_size,
        heads = num_kv_heads,
        expand_factor = mlp_expand_factor
    )
model=Transformer(
    num_tokens = vocab_size,
    dim = hidden_size,
    depth = num_hidden_layers,
    heads = num_heads,
    dim_head = dim_head,
    kv_heads = num_kv_heads,
    use_sparse_attn = True,
    use_flex_sliding_window = True,
    use_triton_fine_selection = False,
    use_flex_fine_selection = False,
    sparse_attn_kwargs = dict(
        sliding_window_size = sliding_window_size,
        compress_block_size = compress_block_size,
        compress_block_overlap_len = overlap_size,
        compress_mlp = compress_mlp,
        selection_block_size = fine_block_size,
        num_selected_blocks = num_fine_selected,
        use_diff_topk = True,
        interpolated_importance_score = False,
        query_heads_share_selected_kv = True
    ),
)
```

Also, I only interned at Amazon for 2 summers. I'm not working there now. I'm working on a startup trying to make better AI agents using a memory mechanism. Modifying NSA is one of the approaches I am considering for this.

**Comment by lucidrains** [comment]
@Pasewark thanks Eric! I'll get the overlapping edge case fixed tomorrow

**Comment by lucidrains** [comment]
@Pasewark should work now

**Comment by Pasewark** [comment]
It works with the overlapping set not equal to 0 now. It still does not work if the fine_block_size!=compress_block_size. I'm not sure how important this case is. Maybe most people will just set these equal anyways? I should also note that if the fine_block_size is set to smaller than the compress_block_size, then even the generation without cache does not produce the same logits as doing a forward pass on the full context.

**Comment by lucidrains** [comment]
@Pasewark yes, that's right, have not finished building that yet

i'll finish the fine block size > compress block size scenario this morning

**Comment by lucidrains** [comment]
@Pasewark ok try it again for fine block size > compress block size, should work now

**Comment by Pasewark** [comment]
It looks like it works both for overlap_size>0 and fine_block_size>=compress_block_size now. Really great work!
